src/qpx_scan_algo.cpp

```diff
@@ -0,0 +1,940 @@
+    printf("Can't run FE/TE test: unsupported media!\n");
+    return 1;
+  }
+  wait_unit_ready(dev, 6);
+  if (plugin->start_test(CHK_FETE, lba, speed)) {
+    printf("Scan init failed!\n");
+    return 2;
+  }
+  gettimeofday(&s, NULL);
+  //    wait_unit_ready(dev, 6);
+  printf("\nTesting %ld sectors: %ld - %ld\n", lba_end - lba_sta + 1, lba_sta,
+         lba_end);
+  printf("         lba |        speed        |  FE  |  TE\n");
+  gettimeofday(&blks, NULL);
+  for (; (!stop_req) && lba < lba_end;) {
+    lbao = lba;
+  block_retry:
+    if (plugin->scan_block((void *)&ft, &lba)) {
+      printf("\nBlock scan error! terminating...\n");
+      stop_req = 1;
+    }
+    if (lba < 0) {
+      if (retry--)
+        goto block_retry;
+      printf("\nDrive returned negative LBA %d times! terminating...\n",
+             MAX_RETRY);
+      stop_req = 1;
+    }
+    if (lba == lbao) {
+      if (retry--)
+        goto block_retry;
+      printf("\nDrive returned same LBA %d times! terminating...\n", MAX_RETRY);
+      stop_req = 1;
+    }
+    retry = MAX_RETRY;
+    gettimeofday(&blke, NULL);
+    ft_max.EMAX(ft);
+    calc_cur_speed(lba - lbao);
+    //		 show current data
+    printf("cur : %6ld | %6.2f X %5d kB/s | %4d | %4d\n", lba, spdX, spdKB,
+           ft.fe, ft.te);
+    blks.tv_sec = blke.tv_sec;
+    blks.tv_usec = blke.tv_usec;
+#ifdef USE_FFLUSH
+    fflush(stdout);
+#endif
+  }
+  plugin->end_test();
+  gettimeofday(&e, NULL);
+  show_avg_speed(lba);
+  printf("\n%ld sectors tested: %ld - %ld\n", lba - lba_sta, lba_sta, lba - 1);
+  printf("Test summary:\n");
+  printf("       FE  |  TE\n");
+  printf("max : %4d | %4d\n", ft_max.fe, ft_max.te);
+#ifdef USE_FFLUSH
+  fflush(stdout);
+#endif
+  return 0;
+}
+
+int qscanner::run_dvd_ta() {
+  struct cdvd_ta ta;
+  long lba;
+  if (!attached)
+    return -1;
+  if (!(dev->media.type & DISC_DVD))
+    return 1;
+  printf("Running DVD Time Analyser test...\n");
+  if (plugin->start_test(CHK_TA, lba, speed)) {
+    printf("Scan init failed!\n");
+    return 2;
+  }
+
+  for (int i = 0; i < 3 * dev->media.layers; i++) {
+    ta.pass = i;
+    plugin->scan_block(&ta, &lba);
+    printf("   idx    pits   lands\n");
+    for (int idx = 0; idx < TA_HIST_SIZE; idx++) {
+      printf("TA %3d  %6d  %6d\r", idx, ta.pit[idx], ta.land[idx]);
+    }
+  }
+
+  // not implemented
+  return -1;
+}
+
+//
+//  BD algo's
+//
+
+int qscanner::run_bd_errc() {
+  bd_errc err, err_tot, err_max;
+  int errc_data;
+  long lba = lba_sta;
+  long lbas;
+  long lbao;
+  if (!attached)
+    return -1;
+  if (!(dev->media.type & DISC_BD))
+    return 1;
+  //    lba=0; slba=0;
+  errc_data = plugin->errc_data();
+  //    seek(dev,lba);
+  if (plugin->start_test(CHK_ERRC_BD, lba, speed)) {
+    printf("BD ERRC test init failed!\n");
+    return 2;
+  }
+  printf("Running BD Error Correction test at speed %d...\n", speed);
+
+  spd1X = 4500;
+  gettimeofday(&s, NULL);
+  wait_unit_ready(dev, 6);
+  lbas = lba;
```

Comment:
Codacy found an issue: [Variable 'lbas' is assigned a value that is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)